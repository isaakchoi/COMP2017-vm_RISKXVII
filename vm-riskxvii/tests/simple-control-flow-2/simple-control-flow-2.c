// Isaak Choi
// 520488399
// icho6322

// Constants
#define READ_INT_ADDR   (0x0816)
#define WRITE_UINT_ADDR (0x0808)
#define WRITE_CHAR_ADDR (0x0800)
#define MALLOC_ADDR     (0x0830)
#define FREE_ADDR       (0x0834)
#define READ_CHAR_ADDR  (0x0812)
#define WRITE_INT_ADDR  (0x0804)

// Other functions
void print_string(char *str) {
    while (*str) *((char *)WRITE_CHAR_ADDR) = *(str++);
}

// Main
int main(){

    // Set hard coded values
    int a = 1;
    int b = 0;
    int c = -1;
    int result;

    // beq
    if (a == 1) {
        print_string("beq (1 == 1): true");
    } else {
        print_string("beq (1 == 1): false");
    }
    print_string("\n");
    if (a == 0) {
        print_string("beq (1 == 0): true");
    } else {
        print_string("beq (1 == 0): false");
    }
    print_string("\n");

    // bne
    if (a != 1) {
        print_string("bne (1 != 1): true");
    } else {
        print_string("bne (1 != 1): false");
    }
    print_string("\n");
    if (a != 0) {
        print_string("bne (1 != 0): true");
    } else {
        print_string("bne (1 != 0): false");
    }
    print_string("\n");

    // blt
    if (a < 1) {
        print_string("blt (1 < 1): true");
    } else {
        print_string("blt (1 < 1): false");
    }
    print_string("\n");
    if (a < 2) {
        print_string("blt (1 < 2): true");
    } else {
        print_string("blt (1 < 2): false");
    }
    print_string("\n");
    if (a < 0) {
        print_string("blt (1 < 0): true");
    } else {
        print_string("blt (1 < 0): false");
    }
    print_string("\n");

    // bltu
    unsigned int ua = 1;
    if (ua < 0) {
        print_string("bltu (1 < 0): true");
    } else {
        print_string("bltu (1 < 0): false");
    }
    print_string("\n");
    if (ua < 1) {
        print_string("bltu (1 < 1): true");
    } else {
        print_string("bltu (1 < 1): false");
    }
    print_string("\n");
    if (ua < 2) {
        print_string("bltu (1 < 2): true");
    } else {
        print_string("bltu (1 < 2): false");
    }
    print_string("\n");

    // bge
    if (a >= 0) {
        print_string("bge (1 >= 0): true");
    } else {
        print_string("bge (1 >= 0): false");
    }
    print_string("\n");
    if (a >= 1) {
        print_string("bge (1 >= 1): true");
    } else {
        print_string("bge (1 >= 1): false");
    }
    print_string("\n");
    if (a >= 2) {
        print_string("bge (1 >= 2): true");
    } else {
        print_string("bge (1 >= 2): false");
    }
    print_string("\n");

    // bgeu
    if (ua >= 0) {
        print_string("bgeu (1 >= 0): true");
    } else {
        print_string("bgeu (1 >= 0): false");
    }
    print_string("\n");
    if (ua >= 1) {
        print_string("bgeu (1 >= 1): true");
    } else {
        print_string("bgeu (1 >= 1): false");
    }
    print_string("\n");
    if (ua >= 2) {
        print_string("bgeu (1 >= 2): true");
    } else {
        print_string("bgeu (1 >= 2): false");
    }
    print_string("\n");

    // Return
    return 0;
}

