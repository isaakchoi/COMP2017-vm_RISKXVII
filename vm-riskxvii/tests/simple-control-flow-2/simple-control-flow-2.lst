
simple-control-flow-2/simple-control-flow-2:     file format elf32-littleriscv


Disassembly of section .text:

00000000 <_start>:
   0:	7ff00113          	li	sp,2047
   4:	05c000ef          	jal	ra,60 <main>
   8:	000017b7          	lui	a5,0x1
   c:	80078623          	sb	zero,-2036(a5) # 80c <ADDR_END+0xc>

00000010 <print_string>:
  10:	fe010113          	addi	sp,sp,-32
  14:	00812e23          	sw	s0,28(sp)
  18:	02010413          	addi	s0,sp,32
  1c:	fea42623          	sw	a0,-20(s0)
  20:	0200006f          	j	40 <print_string+0x30>
  24:	fec42783          	lw	a5,-20(s0)
  28:	00178713          	addi	a4,a5,1
  2c:	fee42623          	sw	a4,-20(s0)
  30:	00001737          	lui	a4,0x1
  34:	80070713          	addi	a4,a4,-2048 # 800 <ADDR_END>
  38:	0007c783          	lbu	a5,0(a5)
  3c:	00f70023          	sb	a5,0(a4)
  40:	fec42783          	lw	a5,-20(s0)
  44:	0007c783          	lbu	a5,0(a5)
  48:	fc079ee3          	bnez	a5,24 <print_string+0x14>
  4c:	00000013          	nop
  50:	00000013          	nop
  54:	01c12403          	lw	s0,28(sp)
  58:	02010113          	addi	sp,sp,32
  5c:	00008067          	ret

00000060 <main>:
  60:	fe010113          	addi	sp,sp,-32
  64:	00112e23          	sw	ra,28(sp)
  68:	00812c23          	sw	s0,24(sp)
  6c:	02010413          	addi	s0,sp,32
  70:	00100793          	li	a5,1
  74:	fef42623          	sw	a5,-20(s0)
  78:	fe042423          	sw	zero,-24(s0)
  7c:	fff00793          	li	a5,-1
  80:	fef42223          	sw	a5,-28(s0)
  84:	fec42703          	lw	a4,-20(s0)
  88:	00100793          	li	a5,1
  8c:	00f71863          	bne	a4,a5,9c <main+0x3c>
  90:	40000513          	li	a0,1024
  94:	f7dff0ef          	jal	ra,10 <print_string>
  98:	00c0006f          	j	a4 <main+0x44>
  9c:	41400513          	li	a0,1044
  a0:	f71ff0ef          	jal	ra,10 <print_string>
  a4:	42800513          	li	a0,1064
  a8:	f69ff0ef          	jal	ra,10 <print_string>
  ac:	fec42783          	lw	a5,-20(s0)
  b0:	00079863          	bnez	a5,c0 <main+0x60>
  b4:	42c00513          	li	a0,1068
  b8:	f59ff0ef          	jal	ra,10 <print_string>
  bc:	00c0006f          	j	c8 <main+0x68>
  c0:	44000513          	li	a0,1088
  c4:	f4dff0ef          	jal	ra,10 <print_string>
  c8:	42800513          	li	a0,1064
  cc:	f45ff0ef          	jal	ra,10 <print_string>
  d0:	fec42703          	lw	a4,-20(s0)
  d4:	00100793          	li	a5,1
  d8:	00f70863          	beq	a4,a5,e8 <main+0x88>
  dc:	45400513          	li	a0,1108
  e0:	f31ff0ef          	jal	ra,10 <print_string>
  e4:	00c0006f          	j	f0 <main+0x90>
  e8:	46800513          	li	a0,1128
  ec:	f25ff0ef          	jal	ra,10 <print_string>
  f0:	42800513          	li	a0,1064
  f4:	f1dff0ef          	jal	ra,10 <print_string>
  f8:	fec42783          	lw	a5,-20(s0)
  fc:	00078863          	beqz	a5,10c <main+0xac>
 100:	47c00513          	li	a0,1148
 104:	f0dff0ef          	jal	ra,10 <print_string>
 108:	00c0006f          	j	114 <main+0xb4>
 10c:	49000513          	li	a0,1168
 110:	f01ff0ef          	jal	ra,10 <print_string>
 114:	42800513          	li	a0,1064
 118:	ef9ff0ef          	jal	ra,10 <print_string>
 11c:	fec42783          	lw	a5,-20(s0)
 120:	00f04863          	bgtz	a5,130 <main+0xd0>
 124:	4a400513          	li	a0,1188
 128:	ee9ff0ef          	jal	ra,10 <print_string>
 12c:	00c0006f          	j	138 <main+0xd8>
 130:	4b800513          	li	a0,1208
 134:	eddff0ef          	jal	ra,10 <print_string>
 138:	42800513          	li	a0,1064
 13c:	ed5ff0ef          	jal	ra,10 <print_string>
 140:	fec42703          	lw	a4,-20(s0)
 144:	00100793          	li	a5,1
 148:	00e7c863          	blt	a5,a4,158 <main+0xf8>
 14c:	4cc00513          	li	a0,1228
 150:	ec1ff0ef          	jal	ra,10 <print_string>
 154:	00c0006f          	j	160 <main+0x100>
 158:	4e000513          	li	a0,1248
 15c:	eb5ff0ef          	jal	ra,10 <print_string>
 160:	42800513          	li	a0,1064
 164:	eadff0ef          	jal	ra,10 <print_string>
 168:	fec42783          	lw	a5,-20(s0)
 16c:	0007d863          	bgez	a5,17c <main+0x11c>
 170:	4f400513          	li	a0,1268
 174:	e9dff0ef          	jal	ra,10 <print_string>
 178:	00c0006f          	j	184 <main+0x124>
 17c:	50800513          	li	a0,1288
 180:	e91ff0ef          	jal	ra,10 <print_string>
 184:	42800513          	li	a0,1064
 188:	e89ff0ef          	jal	ra,10 <print_string>
 18c:	00100793          	li	a5,1
 190:	fef42023          	sw	a5,-32(s0)
 194:	51c00513          	li	a0,1308
 198:	e79ff0ef          	jal	ra,10 <print_string>
 19c:	42800513          	li	a0,1064
 1a0:	e71ff0ef          	jal	ra,10 <print_string>
 1a4:	fe042783          	lw	a5,-32(s0)
 1a8:	00079863          	bnez	a5,1b8 <main+0x158>
 1ac:	53000513          	li	a0,1328
 1b0:	e61ff0ef          	jal	ra,10 <print_string>
 1b4:	00c0006f          	j	1c0 <main+0x160>
 1b8:	54400513          	li	a0,1348
 1bc:	e55ff0ef          	jal	ra,10 <print_string>
 1c0:	42800513          	li	a0,1064
 1c4:	e4dff0ef          	jal	ra,10 <print_string>
 1c8:	fe042703          	lw	a4,-32(s0)
 1cc:	00100793          	li	a5,1
 1d0:	00e7e863          	bltu	a5,a4,1e0 <main+0x180>
 1d4:	55800513          	li	a0,1368
 1d8:	e39ff0ef          	jal	ra,10 <print_string>
 1dc:	00c0006f          	j	1e8 <main+0x188>
 1e0:	56c00513          	li	a0,1388
 1e4:	e2dff0ef          	jal	ra,10 <print_string>
 1e8:	42800513          	li	a0,1064
 1ec:	e25ff0ef          	jal	ra,10 <print_string>
 1f0:	fec42783          	lw	a5,-20(s0)
 1f4:	0007c863          	bltz	a5,204 <main+0x1a4>
 1f8:	58000513          	li	a0,1408
 1fc:	e15ff0ef          	jal	ra,10 <print_string>
 200:	00c0006f          	j	20c <main+0x1ac>
 204:	59400513          	li	a0,1428
 208:	e09ff0ef          	jal	ra,10 <print_string>
 20c:	42800513          	li	a0,1064
 210:	e01ff0ef          	jal	ra,10 <print_string>
 214:	fec42783          	lw	a5,-20(s0)
 218:	00f05863          	blez	a5,228 <main+0x1c8>
 21c:	5a800513          	li	a0,1448
 220:	df1ff0ef          	jal	ra,10 <print_string>
 224:	00c0006f          	j	230 <main+0x1d0>
 228:	5bc00513          	li	a0,1468
 22c:	de5ff0ef          	jal	ra,10 <print_string>
 230:	42800513          	li	a0,1064
 234:	dddff0ef          	jal	ra,10 <print_string>
 238:	fec42703          	lw	a4,-20(s0)
 23c:	00100793          	li	a5,1
 240:	00e7d863          	bge	a5,a4,250 <main+0x1f0>
 244:	5d000513          	li	a0,1488
 248:	dc9ff0ef          	jal	ra,10 <print_string>
 24c:	00c0006f          	j	258 <main+0x1f8>
 250:	5e400513          	li	a0,1508
 254:	dbdff0ef          	jal	ra,10 <print_string>
 258:	42800513          	li	a0,1064
 25c:	db5ff0ef          	jal	ra,10 <print_string>
 260:	5f800513          	li	a0,1528
 264:	dadff0ef          	jal	ra,10 <print_string>
 268:	42800513          	li	a0,1064
 26c:	da5ff0ef          	jal	ra,10 <print_string>
 270:	fe042783          	lw	a5,-32(s0)
 274:	00078863          	beqz	a5,284 <main+0x224>
 278:	60c00513          	li	a0,1548
 27c:	d95ff0ef          	jal	ra,10 <print_string>
 280:	00c0006f          	j	28c <main+0x22c>
 284:	62000513          	li	a0,1568
 288:	d89ff0ef          	jal	ra,10 <print_string>
 28c:	42800513          	li	a0,1064
 290:	d81ff0ef          	jal	ra,10 <print_string>
 294:	fe042703          	lw	a4,-32(s0)
 298:	00100793          	li	a5,1
 29c:	00e7f863          	bgeu	a5,a4,2ac <main+0x24c>
 2a0:	63800513          	li	a0,1592
 2a4:	d6dff0ef          	jal	ra,10 <print_string>
 2a8:	00c0006f          	j	2b4 <main+0x254>
 2ac:	64c00513          	li	a0,1612
 2b0:	d61ff0ef          	jal	ra,10 <print_string>
 2b4:	42800513          	li	a0,1064
 2b8:	d59ff0ef          	jal	ra,10 <print_string>
 2bc:	00000793          	li	a5,0
 2c0:	00078513          	mv	a0,a5
 2c4:	01c12083          	lw	ra,28(sp)
 2c8:	01812403          	lw	s0,24(sp)
 2cc:	02010113          	addi	sp,sp,32
 2d0:	00008067          	ret
