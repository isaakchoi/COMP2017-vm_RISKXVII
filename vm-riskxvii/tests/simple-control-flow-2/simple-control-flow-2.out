beq (1 == 1): true
beq (1 == 0): false
bne (1 != 1): false
bne (1 != 0): true
blt (1 < 1): false
blt (1 < 2): true
blt (1 < 0): false
bltu (1 < 0): false
bltu (1 < 1): false
bltu (1 < 2): true
bge (1 >= 0): true
bge (1 >= 1): true
bge (1 >= 2): false
bgeu (1 >= 0): true
bgeu (1 >= 1): true
bgeu (1 >= 2): false
CPU Halt Requested
